service-base/src/main/java/org/eclipse/hono/service/metric/DeviceConnectionTimeTracker.java

```diff
@@ -0,0 +1,167 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.service.metric;
+
+import io.vertx.core.TimeoutStream;
+import io.vertx.core.Vertx;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+/**
+ * TODO.
+ */
+public final class DeviceConnectionTimeTracker {
+
+    private static final long DEVICE_CONNECTION_TIME_RECORDING_INTERVAL_IN_MS = TimeUnit.SECONDS.toMillis(10);
+    protected final Logger log = LoggerFactory.getLogger(DeviceConnectionTimeTracker.class);
+    private final AtomicLong noOfDeviceConnections;
+    private final Consumer<Long> recorder;
+    private final AtomicReference<Instant> startInstant = new AtomicReference<>();
+    private final String tenantId;
+    private final TimeoutStream timer;
+
+    private DeviceConnectionTimeTracker(
+            final String tenantId, final Vertx vertx,
+            final long noOfDeviceConnections,
+            final Consumer<Long> recorder) {
+        this.tenantId = tenantId;
+        this.noOfDeviceConnections = new AtomicLong(noOfDeviceConnections);
+        this.recorder = recorder;
+        this.startInstant.set(Instant.now());
+        this.timer = vertx
+                .periodicStream(DEVICE_CONNECTION_TIME_RECORDING_INTERVAL_IN_MS)
```

Comment:
I do not quite see why you want to use a periodicStream here. Can you explain?

Reply:
Here I use the periodicStream mainly to take advantage of its methods namely `pause()` and `resume()`. Let us assume a new device is connected or disconnected, then the `updateNoOfDeviceConnections(...)` is called. This method reports the connection time before updating the number of device connections with the actual value. While it reports the connection time, it is also possible that the timer reporting at regular intervals also reports at the same time. In order to avoid that, the timer is paused, the value is reported and then finally the timer is resumed.

Reply:
IMHO that cannot happen because both handlers will be run on the same event loop, i.e. while the handler executing *updateNoOfDeviceConnections* is running, vert.x will make sure that the handler triggered by the timer is not run concurrently. In fact, vert.x guarantees that only exactly one handler will run at any given time on an event loop. But may be you should verify by checking if the timer task is run on the same thread as the handler updating the number of devices. But I am pretty sure that it is ...

Reply:
As I checked, the timer task is running on the same thread as you mentioned above. Then in this case, I will simply use `vertx.setPeriodic(...)` instead of a _periodicStream_.

Reply:
I think that will make it much easier to read/understand

Reply:
@sophokles73 I have pushed the corresponding changes. Would you mind taking a look?